 <div class="alert alert-success" style="border-color: green; border-radius: 5px">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>

Денис, доброго дня! рад приветствовать тебя) <br />
<br />
Меня зовут Николай Шавлюго. <br />И на этом этапе твоего движения к новой профессии от меня будут комментарии по написанному коду. <br />Чтобы меньше "мусорить эфир" и чтобы максимально наглядно отделяться от основного кода, есть предложение условиться в некоторых обозначениях:<br /> <br />
<div class="alert alert-success" style="border-color: green; border-radius: 5px">
    <p><u><b>✅ КОММЕНТАРИЙ РЕВЬЮЕРА</b></u></p>
    <p>Всё хорошо.</p>
</div>

<div class="alert alert-warning" style="border-color: darkorange; border-radius: 5px">
    <p><u><b>⚠️ КОММЕНТАРИЙ РЕВЬЮЕРА</b></u></p>
    <p>Обрати внимание, есть потенциал</p>
</div>

<div class="alert alert-danger" style="border-color: darkred; border-radius: 5px">
    <p><u><b>❌ КОММЕНТАРИЙ РЕВЬЮЕРА</b></u></p>
    <p>Нужно ещё раз глянуть</p>
</div>
</font><br /><br />
<font color='green'>Моей целью является не "уличить" в не знании, а просто высказать сверху твоих знаний - свой опыт, что бы тебе использовать его для своего дальнейшего успеха) И очень здорово будет, если удастся задавать вопросы, да и вообще - всячески доставать меня, если я по каким-то причинам не приму проект:) При этом, из своего опыта скажу, хорошо и важно, когда переписка ревьюера и студента - сохраняется на следующие проекты и даже на будущую практическую деятельность.<br>
<br> Я сразу же начал общаться на ты:) но если это не очень удобно, то дай знать)) <BR> В ПУТЬ!<br></div>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подключим-необходимые-модули-и-библиотеки-Python" data-toc-modified-id="Подключим-необходимые-модули-и-библиотеки-Python-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подключим необходимые модули и библиотеки Python</a></span></li><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Предварительный-осмотр-данных" data-toc-modified-id="Предварительный-осмотр-данных-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Предварительный осмотр данных</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Алгоритм преобразования</a></span><ul class="toc-item"><li><span><a href="#Алгоритм" data-toc-modified-id="Алгоритм-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Алгоритм</a></span></li><li><span><a href="#Обоснование" data-toc-modified-id="Обоснование-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Обоснование</a></span></li></ul></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Проверка алгоритма</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Подключим необходимые модули и библиотеки Python

In [1]:
# подключим модули
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

# создадим константу для фиксации псевдослучайности
STATE = 777

## Загрузка данных
Данные предоставлены в виде csv-файла.

In [2]:
insurance = pd.read_csv('/datasets/insurance.csv')

### Предварительный осмотр данных
- Посмотрим первые 5 строк таблицы,
- Основные статистики количественных переменных,
- Сводную информацию о таблице.

In [3]:
print('*'*20,'Первые пять строк таблицы.','*'*20)
display(insurance.head())
print('*'*10,'Основные статистики количественных переменных.','*'*10)
display(insurance.describe())
print('*'*19,'Сводная информация о таблице.','*'*19)
display(insurance.info())

******************** Первые пять строк таблицы. ********************


Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

********** Основные статистики количественных переменных. **********


Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

******************* Сводная информация о таблице. *******************
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


None

<div class="alert alert-success" style="border-color: green; border-radius: 5px">
    <p><u><b>✅ КОММЕНТАРИЙ РЕВЬЮЕРА</b></u></p>
Данные на базе) ОК, хорошо, когда есть аналитическая культура - обязательно делать первичный анализ данных, не смотря на проект)

## Умножение матриц

Проверим как влияет на качество предсказаний линейной регрессии умножение матрицы признаков на произвольную обратимую матрицу.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $a$ — вектор предсказаний

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

- $quality$ — качество модели

Предсказания:

$$
a = Xw \qquad (1)
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y) \qquad (2)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y \qquad (3)
$$

Качество модели определим как разность между векторами предсказания и целевого признака: 

$$
quality = a - y \qquad (4)
$$

Перепишем формулы (1), (3) и (4) для случая домножения признаков на обратимую матрицу:

$$
a' = XPw' \qquad (1')
$$

$$
w' = ((XP)^T XP)^{-1} (XP)^T y \qquad (3')
$$

$$
quality' = a' - y \qquad (4')
$$

Предположим, что качество не должно изменится, следовательно:

$$
quality' = quality
$$

Произведём соответствующие подстановки:

$$
a' - y = a - y
$$

$$
XP((XP)^T XP)^{-1} (XP)^T y - y = X(X^T X)^{-1} X^T y - y
$$

Применим свойство матриц (так нельзя, правые матрицы не квадратные):

$$
(AB)^{-1} = B^{-1}A^{-1}
$$

$$
(XP)(XP)^{-1}((XP)^T)^{-1} (XP)^T y - y = XX^{-1}(X^T)^{-1} X^T y - y
$$

Применим свойство матриц:

$$
AA^{-1} = A^{-1}A = E
$$

$$
EE y - y = EE y - y
$$

Равенство докуазано. 

**При умножении признаков на обратимую матрицу, качество линейной регрессии не меняется.**

<div class="alert alert-danger" style="border-color: darkred; border-radius: 5px">
    <p><u><b>❌ КОММЕНТАРИЙ РЕВЬЮЕРА</b></u></p>
У матриц есть такие свойста<br>
- ($A(BC) = (AB)C$); это свойство ассоциативности (сочетательности) матриц<br>
- $(P^T)^{-1} = (P^{-1})^T$; то есть: транспортированная оратная матрица есть обратная матрица транспортированная <br>
- $(AB)^T = B^T A^T$; то есть: транспонированное произведение матриц равно произведению транспонированных матриц <br>
- $(AB)^{-1} = B^{-1} A^{-1}$,то есть: обратная матрица от произведения двух матриц равна обратной матрицы B(второй) к обратной матрице А(первой)<br>
- $P P^{-1} = E$ ; умножение матрицы на свою обратную - есть единичная матрица.<br>
- $(ABC)^{-1} =C^{-1} B^{-1} A^{-1}$<br>
    
Посмотри, пожалуйста, доказательство - мне кажется, что некоторые свойства не соблюдаются..<br>
    
Идейно - всё верно. Через выражение при подстановке $w' = ((XP)^T XP)^{-1} (XP)^T y$) надо дойти до $w' = P^{-1}w$ , где $w = (X^T X)^{-1} X^T y$ Это и будет доказательством. При этом важно!!!! $ (X^T X)^{-1}$ - не надо раскрывать, так как для неквадратных матриц ($X$ и $X^T$) обратных матриц не существует

$$
a = Xw = XEw = XPP^{-1}w = (XP)P^{-1}w = (XP)w'
$$

$$
w' = ((XP)^T XP)^{-1} (XP)^T y
$$
$$
w' = (P^T (X^T X) P)^{-1} (XP)^T y
$$
$$
w' = P^{-1}(X^T X)^{-1}(P^T)^{-1} P^TX^T y
$$
применим подставновку $ E = (P^T)^{-1} P^T $
$$
w' = P^{-1}(X^T X)^{-1}EX^T y
$$
$$
w' = P^{-1}(X^T X)^{-1}X^T y
$$
применим подставновку $ w = (X^T X)^{-1} X^T y $
$$
w' = P^{-1}w
$$

<div class="alert alert-info" style="border-color: blue; border-radius: 5px">
    <p><u><b>КОММЕНТАРИЙ студента</b></u></p>
Спасибо за пояснение, ошибку с применением обратной операции для не квадратной матрицы осознал. Применил формулы, довёл доказательство до конца.

<div class="alert alert-success" style="border-color: green; border-radius: 5px">
    <p><u><b>✅ КОММЕНТАРИЙ РЕВЬЮЕРА 2</b></u></p>
ОК, Денис! Всё увидел, принято.<br>
    
Ещё несколько идей в целом по теме: если рассматривать передачу данных третьим лицам, то к преимуществам данного подхода можно отнести, в первую очередь, ощутимую экономию ресурсов по сравнению с некоторыми известными алгоритмами шифрования, в особенности это касается ассиметричных алгоритмов с достаточно длинными ключами (AES, RSA), которые требуют много процессорного времени.<br><br>
    
К главному недостатку подхода можно отнести увеличение сложности проверки на вырожденность преобразования с ростом количества признаков в выборке. В рассмотренном случае необходимо подобрать подходящую матрицу размера  4×4 . Однако, если признаки будут исчисляться тысячами, что в теории всегда надо положить, то расчет соответствующих определителей может быть крайне ресурсозатратным. Но и эту проблему можно будет решить)<br><br>
    
Если есть время и желание, то можно воту этот маткриал просмотреть:<br>
Обзоры по применению криптографии в машинном обучении: https://arxiv.org/pdf/1902.04109.pdf, https://www.researchgate.net/publication/342746786_Some_Applications_of_Machine_Learning_in_Cryptography
Подробнее про аналогичный алгоритм: https://www.cs.cmu.edu/~rjhall/JOS_revised_May_31a.pdf<br><br>
<font color='green'><font color='green'>
Спасибо за отклик и за проект) <br>
Успешного дальнейшего обучения

## Алгоритм преобразования

### Алгоритм

Для шифрования данных применим домножение матрицы признаков на случайно сгенерированную обратимую матрицу размерностью 4x4 (4-число признаков). 
1) Создадим случайно сгенерированную матрицу с фиксированным значением RandomState,

2) Проверим, что полученная матрица является обратимой,

3) Изменим матрицу признаков, умножив её на обратимую матрицу,

4) Обучим модель регрессии,

5) Проведём предсказание на признаках, также прошедших изменение.

<div class="alert alert-success" style="border-color: green; border-radius: 5px">
    <p><u><b>✅ КОММЕНТАРИЙ РЕВЬЮЕРА</b></u></p>
да, тут как раз и нужен план действий для пункта 4) Делай раз, делай два, делай три..

### Обоснование

Преобразование матрицы признаков при помощи домножения на обратимую матрицу не должно изменять качество модели, что было доказано в пункте "Умножение матриц".

## Проверка алгоритма

Выделим из данных признаки и целевой признак.

In [4]:
features = insurance.drop('Страховые выплаты',axis=1)
target = insurance['Страховые выплаты']

Разабьем данные на обучающую и тестовую части в отношении 75/25.

In [5]:
features_train, features_test, target_train, target_test = train_test_split(
    features, target, test_size=0.25, random_state=STATE)

Обучим модель линейной регрессии на нешифрованных данных.

In [6]:
model_before = LinearRegression()
model_before.fit(features_train, target_train)

LinearRegression()

Вычислим метрику R2 для нешифрованных данных.

In [7]:
R2_before = r2_score(target_test, model_before.predict(features_test))
print("Значение метрики R2 до шифрования признаков", R2_before)

Значение метрики R2 до шифрования признаков 0.42282460954817036


Создадим обратимую матрицу 4x4 со случайными значениями, проверим что детерминант матрицы не равен нулю это гарантирует нам обратимость матрицы.

In [8]:
np.random.seed(STATE)
det = 0
while det == 0:
    coding_matrix = np.random.randint(1, 10, (4,4))
    det = np.linalg.det(coding_matrix)
coding_matrix

array([[8, 7, 8, 2],
       [8, 5, 8, 9],
       [8, 3, 1, 2],
       [3, 5, 6, 8]])

<div class="alert alert-success" style="border-color: green; border-radius: 5px">
    <p><u><b>✅ КОММЕНТАРИЙ РЕВЬЮЕРА</b></u></p>
ок, это важно - проверять на обратимость!

Умножим признаки на обратимую матрицу, получим зашифрованные наборы признаков.

In [9]:
features_train_coding = features_train @ coding_matrix
features_test_coding = features_test @ coding_matrix

Обучим модель на шифрованных признаках.

In [10]:
model_coding = LinearRegression()
model_coding.fit(features_train_coding, target_train)

LinearRegression()

Вычислим метрику R2 для шифрованных данных.

In [11]:
R2_coding = r2_score(target_test, model_coding.predict(features_test_coding))
print("Значение метрики R2 после шифрования признаков", R2_coding)

Значение метрики R2 после шифрования признаков 0.42282460954812395


Как видно метрика R2 = 0.4228 не изменила своего значения.

## Вывод

В ходе проекта было получено теоритическое обоснование возможности шифрования без потери качества предсказания модели. Шифрование проводилось посредством домножения признаков на случайную обратимую матрицу.

Также была проведена практическая проверка сохранения качества модели после шифрования по метрике R2.

<div class="alert alert-success" style="border-color: green; border-radius: 5px">
    <p><u><b>✅ КОММЕНТАРИЙ РЕВЬЮЕРА</b></u></p>
ОК, Денис, чёткая работа! Сама практическая цель достигнута, ОК<br>
Тут всё верно. Есть всего один момент для дополнений:<br>
    
В разделе "Умножение матриц" ожидается алгебраическое доказательство чуть иначе. В тексте постарался описать проблематику.<font color='red'><br>
Посмотри, пожалуйста)<br><font color='green'><font color='green'>
Спасибо) Буду ждать.

<div class="alert alert-info" style="border-color: blue; border-radius: 5px">
    <p><u><b>КОММЕНТАРИЙ студента</b></u></p>
Спасибо за ревью. Работать с формулами матриц нужно аккуратнее и внимательнее )

    Вопрос: часто ли на практике приходится заниматься шифрованием данных при их получении? Кажется это немного излишним и задание сделано просто для знакомства с линейной алгеброй и свойством матриц.

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [ ]  Весь код выполняется без ошибок
- [ ]  Ячейки с кодом расположены в порядке исполнения
- [ ]  Выполнен шаг 1: данные загружены
- [ ]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [ ]  Указан правильный вариант ответа
    - [ ]  Вариант обоснован
- [ ]  Выполнен шаг 3: предложен алгоритм преобразования
    - [ ]  Алгоритм описан
    - [ ]  Алгоритм обоснован
- [ ]  Выполнен шаг 4: алгоритм проверен
    - [ ]  Алгоритм реализован
    - [ ]  Проведено сравнение качества моделей до и после преобразования